---
title: "Estimating mRNA secondary structure stability at 30C with ensemble deltaG, minimum free energy (mfe) deltaG, and mean base-pair probability." 
subtitle: "Full set of genes: 1458 genes, each with SNP(s) in their transcripts (5'UTR, CDS, and 3'UTR).\n\nPartial set of genes: 176 genes, each different at the nucleotide sequence level and identical at the amino acid sequence level across 22 strains"
date: "7/18/2022"
output: html_document
#rerunning anastacia's models with PPR^(1/2) as the response and with deltaGs only scaled by 1000
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, message=FALSE, warning=FALSE}
library(car)
library(lme4)
library(reshape2)
```


#### DeltaG of the thermodynamic ensemble {.tabset .tabset-dropdown}
##### Menu
##### Define the response variable, sqrtPPR.
```{r response_variable0}
response <- read.csv("n_PPR_1458_d.csv")
dataset <- melt(response, id.vars = "gene", variable.name = "strain", value.name = "PPR")
dataset$sqrtPPR <- (dataset$PPR)^(1/2)
```
##### Read in Data
```{r F0}
explanatory0 <- read.csv('ensemble_fe_1458_30C.csv')
explanatory0 <- melt(explanatory0, id.vars = "gene", variable.name = "strain", value.name = "ensembleG")
dataset$ensembleG = explanatory0$ensembleG/1000 #divide by 1000 -> deltaG in Mcal/mol -> don't need to scale slope
```

##### Summary Results of Mixed Model
``` {r F1}
lmm0 <- lmer(sqrtPPR ~ ensembleG + (1+ensembleG|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="Nelder_Mead", optCtrl=list(maxfun=1e5)))
summary(lmm0)
```

##### 95% Confidence Interval of Mixed Model
``` {r F1confint}
confint(lmm0, parm="ensembleG", level=0.95)
```

##### Fixed Effects Slope Significance Results
``` {r F2}
lmm00 <- lmer(sqrtPPR ~ (1+ensembleG|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="Nelder_Mead", optCtrl=list(maxfun=1e5)))
anova(lmm0, lmm00)
```

##### Diagnostic Plots
``` {r F3, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm0)

scatter.smooth(fitted(lmm0), res, col='cornflowerblue', span=0.1, xlab='Fitted Values', ylab='Residuals')
scatter.smooth(dataset$ensembleG, res, col='cornflowerblue', span=0.1, xlab="Ensemble DeltaG Scaled by 1/1000", ylab='Residuals')
qqnorm(res, col='cornflowerblue')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='cornflowerblue')
rm(list = ls(all.names = TRUE))
```





#### DeltaG of the minimum free energy (mfe) structure {.tabset .tabset-dropdown}
##### Menu
##### Define the response variable, sqrtPPR.
```{r response_variable1}
response <- read.csv('n_PPR_1458_d.csv')
dataset <- melt(response, id.vars = "gene", variable.name = "strain", value.name = "PPR")
dataset$sqrtPPR <- (dataset$PPR)^(1/2)
```
##### Read in Data
```{r A0}
explanatory1 <- read.csv('optimal_struc_mfe_1458_30C.csv')
explanatory1 <- melt(explanatory1, id.vars = "gene", variable.name = "strain", value.name = "optimalG")
dataset$optimalG <- explanatory1$optimalG/1000 #divide by 1000 -> deltaG in Mcal/mol -> don't need to scale slope
```

##### Summary Results of Mixed Model
``` {r A1}
lmm1 <- lmer(sqrtPPR ~ optimalG + (1+optimalG|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="Nelder_Mead", optCtrl=list(maxfun=1e5)))
summary(lmm1)
```

##### 95% Confidence Interval of Mixed Model
``` {r A1confint}
confint(lmm1, parm="optimalG", level=0.95)
```

##### Fixed Effects Slope Significance Results
``` {r A2}
lmm01 <- lmer(sqrtPPR ~ (1+optimalG|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="Nelder_Mead", optCtrl=list(maxfun=1e5)))
anova(lmm1, lmm01)
```

##### Diagnostic Plots
``` {r A3, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm1)

scatter.smooth(fitted(lmm1), res, col='cornflowerblue', span=0.1, xlab='Fitted Values', ylab='Residuals')
scatter.smooth(dataset$optimalG, res, col='cornflowerblue', span=0.1, xlab="mfe DeltaG Scaled by 1/1000", ylab='Residuals')
qqnorm(res, col='cornflowerblue')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='cornflowerblue')
rm(list = ls(all.names = TRUE))
```





#### Arithmetic mean gene basepair probability {.tabset .tabset-dropdown}
##### Menu
##### Define the response variable, sqrtPPR.
```{r response_variable2}
response <- read.csv('n_PPR_1458_d.csv')
dataset <- melt(response, id.vars = "gene", variable.name = "strain", value.name = "PPR")
dataset$sqrtPPR <- (dataset$PPR)^(1/2)
```
##### Read in Data
```{r W0}
explanatory2 <- read.csv('mean_gene_bp_prob_1458_30C.csv')
explanatory2 <- melt(explanatory2, id.vars = "gene", variable.name = "strain", value.name = "bp_prob")
dataset$bp_prob <- (explanatory2$bp_prob)*10  #multiplying by 10 makes the model work with no warnings 
#after finding the slope of the model, we multiply the slope times 10 to get the units back where they were
#because rise/run so the bp_prob is in the denominator, so multiplying the slope by 10 undoes the multiplying the bp_prob by 10
```

##### Summary Results of Mixed Model
``` {r W1}
lmm2 <- lmer(sqrtPPR ~ bp_prob + (1+bp_prob|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="Nelder_Mead", optCtrl=list(maxfun=1e5)))
summary(lmm2)
```

##### 95% Confidence Interval of Mixed Model
``` {r W1confint}
confint(lmm2, parm="bp_prob", level=0.95)
```

##### Fixed Effects Slope Significance Results
``` {r W2}
lmm02 <- lmer(sqrtPPR ~ (1+bp_prob|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
anova(lmm2, lmm02)
```

##### Diagnostic Plots
``` {r W3, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm2)

scatter.smooth(fitted(lmm2), res, col='cornflowerblue', span=0.1, xlab='Fitted Values', ylab='Residuals')
scatter.smooth(dataset$bp_prob, res, col='cornflowerblue', span=0.1, xlab="Mean Gene Basepair Probability", ylab='Residuals')
qqnorm(res, col='cornflowerblue')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='cornflowerblue')
rm(list = ls(all.names = TRUE))
```





### Analysis of mRNA Ssecondary structure on Partial Set of Genes
#### DeltaG of the thermodynamic ensemble {.tabset .tabset-dropdown}
##### Menu
##### Define the response variable, sqrtPPR.
```{r response_variable0_partial}
response <- read.csv("n_PPR_176_d.csv")
dataset <- melt(response, id.vars = "gene", variable.name = "strain", value.name = "PPR")
dataset$sqrtPPR <- (dataset$PPR)^(1/2)
```
##### Read in Data
```{r F0_partial}
explanatory0 <- read.csv('ensemble_fe_176_30C.csv')
explanatory0 <- melt(explanatory0, id.vars = "gene", variable.name = "strain", value.name = "ensembleG")
dataset$ensembleG = explanatory0$ensembleG/1000 #divide by 1000 -> deltaG in Mcal/mol -> don't need to scale slope
```

##### Summary Results of Mixed Model
``` {r F1_partial}
lmm0 <- lmer(sqrtPPR ~ ensembleG + (1+ensembleG|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="Nelder_Mead", optCtrl=list(maxfun=1e5)))
summary(lmm0)
```

##### 95% Confidence Interval of Mixed Model
``` {r F1confint_partial}
confint(lmm0, parm="ensembleG", level=0.95)
```

##### Fixed Effects Slope Significance Results
``` {r F2_partial}
lmm00 <- lmer(sqrtPPR ~ (1+ensembleG|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="Nelder_Mead", optCtrl=list(maxfun=1e5)))
anova(lmm0, lmm00)
```

##### Diagnostic Plots
``` {r F3_partial, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm0)

scatter.smooth(fitted(lmm0), res, col='cornflowerblue', span=0.1, xlab='Fitted Values', ylab='Residuals')
scatter.smooth(dataset$ensembleG, res, col='cornflowerblue', span=0.1, xlab="Ensemble DeltaG Scaled by 1/1000", ylab='Residuals')
qqnorm(res, col='cornflowerblue')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='cornflowerblue')
rm(list = ls(all.names = TRUE))
```





#### DeltaG of the minimum free energy (mfe) structure {.tabset .tabset-dropdown}
##### Menu
##### Define the response variable, sqrtPPR.
```{r response_variable1_partial}
response <- read.csv('n_PPR_176_d.csv')
dataset <- melt(response, id.vars = "gene", variable.name = "strain", value.name = "PPR")
dataset$sqrtPPR <- (dataset$PPR)^(1/2)
```
##### Read in Data
```{r A0_partial}
explanatory1 <- read.csv('optimal_struc_mfe_176_30C.csv')
explanatory1 <- melt(explanatory1, id.vars = "gene", variable.name = "strain", value.name = "optimalG")
dataset$optimalG <- explanatory1$optimalG/1000 #divide by 1000 -> deltaG in Mcal/mol -> don't need to scale slope
```

##### Summary Results of Mixed Model
``` {r A1_partial}
lmm1 <- lmer(sqrtPPR ~ optimalG + (1+optimalG|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="Nelder_Mead", optCtrl=list(maxfun=1e5)))
summary(lmm1)
```

##### 95% Confidence Interval of Mixed Model
``` {r A1confint_partial}
confint(lmm1, parm="optimalG", level=0.95)
```

##### Fixed Effects Slope Significance Results
``` {r A2_partial}
lmm01 <- lmer(sqrtPPR ~ (1+optimalG|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="Nelder_Mead", optCtrl=list(maxfun=1e5)))
anova(lmm1, lmm01)
```

##### Diagnostic Plots
``` {r A3_partial, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm1)

scatter.smooth(fitted(lmm1), res, col='cornflowerblue', span=0.1, xlab='Fitted Values', ylab='Residuals')
scatter.smooth(dataset$optimalG, res, col='cornflowerblue', span=0.1, xlab="mfe DeltaG Scaled by 1/1000", ylab='Residuals')
qqnorm(res, col='cornflowerblue')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='cornflowerblue')
rm(list = ls(all.names = TRUE))
```


#### Arithmetic mean gene basepair probability {.tabset .tabset-dropdown}
##### Menu
##### Define the response variable, sqrtPPR.
```{r response_variable2_partial}
response <- read.csv('n_PPR_176_d.csv')
dataset <- melt(response, id.vars = "gene", variable.name = "strain", value.name = "PPR")
dataset$sqrtPPR <- (dataset$PPR)^(1/2)
```
##### Read in Data
```{r W0_partial}
explanatory2 <- read.csv('mean_gene_bp_prob_176_30C.csv')
explanatory2 <- melt(explanatory2, id.vars = "gene", variable.name = "strain", value.name = "bp_prob")
dataset$bp_prob <- (explanatory2$bp_prob)*10  #multiplying by 10 makes the model work with no warnings 
#after finding the slope of the model, we multiply the slope times 10 to get the units back where they were
#because rise/run so the bp_prob is in the denominator, so multiplying the slope by 10 undoes the multiplying the bp_prob by 10
```

##### Summary Results of Mixed Model
``` {r W1_partial}
lmm2 <- lmer(sqrtPPR ~ bp_prob + (1+bp_prob|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="Nelder_Mead", optCtrl=list(maxfun=1e5)))
summary(lmm2)
```


##### 95% Confidence Interval of Mixed Model
``` {r W1confint_partial}
confint(lmm2, parm="bp_prob", level=0.95)
```

##### Fixed Effects Slope Significance Results
``` {r W2_partial}
lmm02 <- lmer(sqrtPPR ~ (1+bp_prob|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="Nelder_Mead", optCtrl=list(maxfun=1e5)))
anova(lmm2, lmm02)
```


##### Diagnostic Plots
``` {r W3_partial, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm2)

scatter.smooth(fitted(lmm2), res, col='cornflowerblue', span=0.1, xlab='Fitted Values', ylab='Residuals')
scatter.smooth(dataset$bp_prob, res, col='cornflowerblue', span=0.1, xlab="Mean Gene Basepair Probability", ylab='Residuals')
qqnorm(res, col='cornflowerblue')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='cornflowerblue')
rm(list = ls(all.names = TRUE))
```

