---
title: "Comparing Four Measures of Codon Bias with sqrtPPR response variable"
subtitle: "Full set of genes: 1620 genes, each different at the nucleotide sequence level across 22 strains\n\nPartial set of genes: 185 genes, each different at the nucleotide sequence level and identical at the amino acid sequence level across 22 strains\n\nSplit set of genes: 810 genes in each group (high/low medianPA)"
output:
  html_document: default
  pdf_document: default
date: "03/15/2023"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, message=FALSE}
library(car)
library(lme4)
library(reshape2)
```


### Figure S2A - Analysis of Codon Bias on the Full Set of Genes
Define the response variable.
```{r response_variable}
response <- read.csv('n_PPR_1620_d.csv')
dataset <- melt(response, id.vars = "gene", variable.name = "strain", value.name = "PPR")
dataset$sqrtPPR <- sqrt(dataset$PPR)
```

#### Codon Adaptation Index (CAI) with Carbone Reference Set{.tabset .tabset-dropdown}
##### Menu
##### Read in Data
```{r CAI0}
explanatory0 <- read.csv('CAI_Carbone_1620_YRC_diff_d.csv')
explanatory0 <- melt(explanatory0, id.vars = "gene", variable.name = "strain", value.name = "CAI")
dataset$CAI <- explanatory0$CAI
```

##### Summary Results of Mixed Model
``` {r CAI1}
lmm1 <- lmer(sqrtPPR ~ CAI + (1+CAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
summary(lmm1)
```

##### Fixed Effects Slope Significance Results
``` {r CAI2}
lmm2 <- lmer(sqrtPPR ~ (1+CAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
cai1620<-anova(lmm1, lmm2)
cai1620
```

##### p-value
``` {r CAI1620CI}
cai1620[2,8]
```

##### 95% Confidence Interval of Mixed Model
``` {r CAI4}
confint(lmm1, parm="CAI", level=0.95)
```

##### Diagnostic Plots
``` {r CAI3, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm1)

scatter.smooth(fitted(lmm1), res, col='slateblue', span=0.1, xlab='Fitted Values', ylab='Residuals')
tiff(file="cai_sqrtppr_res_fit.tiff", width = 12, height = 12, units = 'cm', res = 300)
scatter.smooth(fitted(lmm1), res, col='black', span=0.1, xlab='Fitted sqrtPPR values', ylab='Residuals')
dev.off()
scatter.smooth(dataset$CAI, res, col='slateblue', span=0.1, xlab='CAI (Carbone)', ylab='Residuals')
qqnorm(res, col='slateblue')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='slateblue')
```





#### Length-Normalized Codon Adaptation Index (nlCAI) {.tabset .tabset-dropdown}
##### Menu
##### Read in Data
```{r nlCAI0}
explanatory1 <- read.csv('nlCAI_1620_YRC_diff_d.csv')
explanatory1 <- melt(explanatory1, id.vars = "gene", variable.name = "strain", value.name = "nlCAI")
dataset$nlCAI <- explanatory1$nlCAI
```

##### Summary Results of Mixed Model
``` {r nlCAI1}
lmm3 <- lmer(sqrtPPR ~ nlCAI + (1+nlCAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
summary(lmm3)
```

##### Fixed Effects Slope Significance Results
``` {r nlCAI2}
lmm4 <- lmer(sqrtPPR ~ (1+nlCAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
nlcai1620<-anova(lmm3, lmm4)
nlcai1620
```

##### Fixed Effects Slope Significance Results
``` {r nlCAI1620pvalue}
nlcai1620[2,8]
```

##### 95% Confidence Interval of Mixed Model
``` {r nlCAI4}
confint(lmm3, parm="nlCAI", level=0.95)
```

##### Diagnostic Plots
``` {r nlCAI3, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm3)

scatter.smooth(fitted(lmm3), res, col='salmon', span=0.1, xlab='Fitted Values', ylab='Residuals')
scatter.smooth(dataset$nlCAI, res, col='salmon', span=0.1, xlab='nlCAI', ylab='Residuals')
qqnorm(res, col='salmon')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='salmon')
```





#### tRNA Adaptation Index (tAI) {.tabset .tabset-dropdown}
##### Menu
##### Read in Data
```{r tAI0}
explanatory2 <- read.csv('tAI_1620_YRC_diff_d.csv')
explanatory2 <- melt(explanatory2, id.vars = "gene", variable.name = "strain", value.name = "tAI")
dataset$tAI <- explanatory2$tAI
```

##### Summary Results of Model
``` {r tAI1}
lmm5 <- lmer(sqrtPPR ~ tAI + (1+tAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="Nelder_Mead", optCtrl=list(maxfun=1e5)))
summary(lmm5)
```

##### Fixed Effects Slope Significance Results
``` {r tAI2}
lmm6 <- lmer(sqrtPPR ~ (1+tAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
tai1620<-anova(lmm5, lmm6)
tai1620
```

##### p-value
``` {r tAI1620ci}
tai1620[2,8]
```

##### 95% Confidence Interval of Mixed Model
``` {r nltAI4}
confint(lmm5, parm="tAI", level=0.95)
```

##### Diagnostic Plots
``` {r tAI3, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm5)

scatter.smooth(fitted(lmm5), res, col='yellowgreen', span=0.1, xlab='Fitted Values', ylab='Residuals')
scatter.smooth(dataset$tAI, res, col='yellowgreen', span=0.1, xlab='tAI', ylab='Residuals')
qqnorm(res, col='yellowgreen')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='yellowgreen')
```





#### Normalized tRNA Adaptation Index (ntAI) {.tabset .tabset-dropdown}
##### Menu
##### Read in Data
```{r ntAI0}
explanatory3 <- read.csv('ntAI_1620_YRC_diff_d.csv')
explanatory3 <- melt(explanatory3, id.vars = "gene", variable.name = "strain", value.name = "ntAI")
dataset$ntAI <- explanatory3$ntAI
```

##### Summary Results of Model
``` {r ntAI1}
lmm7 <- lmer(sqrtPPR ~ ntAI + (1+ntAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
summary(lmm7)
```

##### Fixed Effects Slope Significance Results
``` {r ntAI2}
lmm8 <- lmer(sqrtPPR ~ (1+ntAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
ntai1620<-anova(lmm7, lmm8)
ntai1620
```

##### p-value
``` {r ntAI1620ci}
ntai1620[2,8]
```

##### 95% Confidence Interval of Mixed Model
``` {r ntAI4}
confint(lmm7, parm="ntAI", level=0.95)
```

##### Diagnostic Plots
``` {r ntAI3, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm7)

scatter.smooth(fitted(lmm7), res, col='peru', span=0.1, xlab='Fitted Values', ylab='Residuals')
scatter.smooth(dataset$ntAI, res, col='peru', span=0.1, xlab='ntAI', ylab='Residuals')
qqnorm(res, col='peru')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='peru')
rm(list=ls())
```




### Figure 2C - With tAI, look at subsets of genes and medianPA as a random effect

### Partial Set of Genes - synonymous only
Define the response variable.
```{r response_variablet}
response <- read.csv('n_PPR_185_d.csv')
dataset <- melt(response, id.vars = "gene", variable.name = "strain", value.name = "PPR")
dataset$sqrtPPR <- sqrt(dataset$PPR)
```

#### tRNA Adaptation Index (tAI) {.tabset .tabset-dropdown}
##### Menu
##### Read in Data
```{r tAIt0}
explanatory0 <- read.csv('tAI_185_YRC_diff_d.csv')
explanatory0 <- melt(explanatory0, id.vars = "gene", variable.name = "strain", value.name = "tAI")
dataset$tAI <- explanatory0$tAI
```

##### Summary Results of Model
``` {r StAIt1}
lmm1 <- lmer(sqrtPPR ~ tAI + (1+tAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
summary(lmm1)
```

##### Fixed Effects Slope Significance Results
``` {r tAIt2}
lmm2 <- lmer(sqrtPPR ~ (1+tAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
tai185<-anova(lmm1, lmm2)
tai185
```

##### p-value
``` {r tAIt185ci}
tai185[2,8]
```

##### 95% Confidence Interval of Mixed Model
``` {r tAIt4}
confint(lmm1, parm="tAI", level=0.95)
```

##### Diagnostic Plots
``` {r StAIt3, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm1)

scatter.smooth(fitted(lmm1), res, col='yellowgreen', span=0.1, xlab='Fitted Values', ylab='Residuals')
scatter.smooth(dataset$tAI, res, col='yellowgreen', span=0.1, xlab='tAI', ylab='Residuals')
qqnorm(res, col='yellowgreen')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='yellowgreen')
```


### High medianPA genes
Define the response variable.
```{r response_variable_highpa}
response <- read.csv('n_PPR_810_highPA_d.csv')
dataset <- melt(response, id.vars = "gene", variable.name = "strain", value.name = "PPR")
dataset$sqrtPPR <- sqrt(dataset$PPR)
```

#### tRNA Adaptation Index (tAI) {.tabset .tabset-dropdown}
##### Menu
##### Read in Data
```{r tAIhighpa0}
explanatory0 <- read.csv('tAI_810_highPA_YRC_diff_d.csv')
explanatory0 <- melt(explanatory0, id.vars = "gene", variable.name = "strain", value.name = "tAI")
dataset$tAI <- explanatory0$tAI
```

##### Summary Results of Model
``` {r tAIhighpa1}
lmm3 <- lmer(sqrtPPR ~ tAI + (1+tAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
summary(lmm3)
```

##### 95% Confidence Interval of Mixed Model
``` {r tAIhighpa}
confint(lmm3, parm="tAI", level=0.95)
```

##### Fixed Effects Slope Significance Results
``` {r tAIhighpa2}
lmm4 <- lmer(sqrtPPR ~ (1+tAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
tai_highpa<-anova(lmm3, lmm4)
tai_highpa
```

``` {r tAIhighpapvalue}
tai_highpa[2,8]
```

##### Diagnostic Plots
``` {r tAIhighpa3, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm3)

scatter.smooth(fitted(lmm3), res, col='yellowgreen', span=0.1, xlab='Fitted Values', ylab='Residuals')
scatter.smooth(dataset$tAI, res, col='yellowgreen', span=0.1, xlab='tAI', ylab='Residuals')
qqnorm(res, col='yellowgreen')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='yellowgreen')
```


### Low medianPA genes
Define the response variable.
```{r response_variable_lowpa}
response <- read.csv('n_PPR_810_lowPA_d.csv')
dataset <- melt(response, id.vars = "gene", variable.name = "strain", value.name = "PPR")
dataset$sqrtPPR <- sqrt(dataset$PPR)
```

#### tRNA Adaptation Index (tAI) {.tabset .tabset-dropdown}
##### Menu
##### Read in Data
```{r tAIlowpa0}
explanatory0 <- read.csv('tAI_810_lowPA_YRC_diff_d.csv')
explanatory0 <- melt(explanatory0, id.vars = "gene", variable.name = "strain", value.name = "tAI")
dataset$tAI <- explanatory0$tAI
```

##### Summary Results of Model
``` {r tAIlowpa1}
lmm5 <- lmer(sqrtPPR ~ tAI + (1+tAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
summary(lmm5)
```

##### 95% Confidence Interval of Mixed Model
``` {r tAIlowpa}
confint(lmm5, parm="tAI", level=0.95)
```

##### Fixed Effects Slope Significance Results
``` {r tAIlowpa2}
lmm6 <- lmer(sqrtPPR ~ (1+tAI|gene), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
tai_810low<-anova(lmm5, lmm6)
tai_810low
```

``` {r tAIlowpapvalue}
tai_810low[2,8]
```

##### Diagnostic Plots
``` {r tAIlowpa3, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm5)

scatter.smooth(fitted(lmm5), res, col='yellowgreen', span=0.1, xlab='Fitted Values', ylab='Residuals')
scatter.smooth(dataset$tAI, res, col='yellowgreen', span=0.1, xlab='tAI', ylab='Residuals')
qqnorm(res, col='yellowgreen')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='yellowgreen')
```



#### Include medianPA as a random effect
#### tRNA Adaptation Index (tAI) {.tabset .tabset-dropdown}
##### Menu
##### Read in Data
Define the response variable.
```{r response_variable_medpa}
response <- read.csv('n_PPR_1620_d.csv')
dataset <- melt(response, id.vars = "gene", variable.name = "strain", value.name = "PPR")
dataset$sqrtPPR <- sqrt(dataset$PPR)
```

```{r tAImedra0}
explanatory0 <- read.csv('tAI_1620_YRC_diff_d.csv')
explanatory0 <- melt(explanatory0, id.vars = "gene", variable.name = "strain", value.name = "tAI")
dataset$tAI <- explanatory0$tAI
```

#include median protein abundance for each gene
```{r pa}
explanatory1 <- read.csv('PA_medians_1620.csv')
explanatory1 <- melt(explanatory1, id.vars = "gene", variable.name = "strain", value.name = "medianPA")
dataset$medianPA = explanatory1$medianPA/10000
```

#check
``` {r check}
head(dataset)
```

##### Summary Results of Model
``` {r tAImedpa1}
lmm7 <- lmer(sqrtPPR ~ tAI + (1+tAI|medianPA), data=dataset, REML = FALSE, control = lmerControl(optimizer="Nelder_Mead", optCtrl=list(maxfun=1e5)))
summary(lmm7)
```

##### Fixed Effects Slope Significance Results
``` {r tAImedpa2}
lmm8 <- lmer(sqrtPPR ~ (1+tAI|medianPA), data=dataset, REML = FALSE, control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=1e5)))
tai1620mpa<-anova(lmm7, lmm8)
tai1620mpa
```

##### p-value
``` {r tAImedpapvalue}
tai1620mpa[2,8]
```

##### 95% Confidence Interval of Mixed Model
``` {r tAImedpaCI}
confint(lmm7, parm="tAI", level=0.95)
```

##### Diagnostic Plots
``` {r tAImedpa3, fig.align='center', fig.height=10, fig.width=10}
par(mfrow=c(2,2))
res = residuals(lmm7)

scatter.smooth(fitted(lmm7), res, col='yellowgreen', span=0.1, xlab='Fitted Values', ylab='Residuals')
scatter.smooth(dataset$tAI, res, col='yellowgreen', span=0.1, xlab='tAI', ylab='Residuals')
qqnorm(res, col='yellowgreen')
qqline(res)
hist(res, main='', xlab='Residuals', ylab='Frequency', col='yellowgreen')
```
